lib/libzpool/uzfs_io.c

```diff
@@ -221,42 +218,67 @@ uzfs_read_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 			bytes = volsize - offset;
 
 		error = dmu_read(os, ZVOL_OBJ, offset, bytes, buf + read, 0);
-		if (error) {
-			ret = UZFS_IO_READ_FAIL;
-			goto exit_with_error;
-		}
+		if (error != 0)
+			goto exit;
+
+		if (md_head != NULL) {
+			blk_metadata_t metadata;
 
-		if ((md != NULL) && (mdlen != NULL)) {
 			/* This assumes metavolblocksize same as volblocksize */
 			get_metaobj_block_details(&metablk, zv, offset);
 
 			mrl = zfs_range_lock(&zv->zv_mrange_lock,
 			    metablk.m_offset, metadatasize, RL_READER);
 			error = dmu_read(os, ZVOL_META_OBJ, metablk.m_offset,
-			    metadatasize, mdata + mread, 0);
-			if (error) {
-				zfs_range_unlock(mrl);
-				ret = UZFS_IO_MREAD_FAIL;
-				goto exit_with_error;
-			}
+			    metadatasize, &metadata, 0);
 			zfs_range_unlock(mrl);
-			mread += metadatasize;
+			if (error != 0)
+				goto exit;
+
+			new_md = NULL;
+			if (md_ent != NULL) {
+				/*
+				 * Join adjacent metadata with the same io
+				 * number into one descriptor. Otherwise
+				 * create a new one.
+				 */
+				if (md_ent->metadata.io_num == metadata.io_num)
+					md_ent->len += bytes;
+				else {
+					new_md = kmem_alloc(
```

Comment:
@jkryl .. I understood the approach of linked list to get the offset and length also. In my ongoing PR, I removed the assumption that metavolblocksize should be same as volblocksize. I can modify read API similar to this which returns array with (metadata and len) [from len, we can calculate offset during socket write]. And, joining of adjacent data with same IO number can done while doing socket write. Would that makes sense?
You can go thru that PR, and drop your comments so that I can modify it according to this requirement.
However, I will start with the approach I mentioned so that you can understand what I meant.

Reply:
just tried to implement as I mentioned above to return array of (metadata, len). but, actually, len won't be different from zv->zv_metavolblocksize (this setting gives the data size at which metadata is stored).
With my ongoing PR, offset & len arguments of read/write APIs should be multiple of zv->zv_metavolblocksize. So, each metadata information (actually IO number currently) is related to datasize of zv->zv_metavolblocksize. With this, we should be able to avoid this linked list.

Reply:
Vitta, as you say, we could remove linked list used internally if we can rely on the fact that all reads and writes are zv_metavolblocksize-aligned, which was the initial reason for introducing linked list. For current over the wire format it is easier to generate it from linked list because we don't need to join adjacent blocks with the same io number into one chunk (although this is easy to do :) ). But regardless, I find the uzfs_read_data API less confusing with linked list instead of array, because the caller does not need to know zv_metavolblocksize in order to assign correct io numbers to data chunks. The only disadvantage I see is the increased number of allocs and frees in case of a read with many data chunks with different IOs inside. For typical case of reading a block with single io number, it is no different from this perspective.

If we are moving from reading/writing data of arbitrary size to more block oriented API (offset and length must be block aligned), then we should perhaps modify the API to take offset in blocks and length in blocks arguments to clearly enforce that at API level. And in that case, I agree that linked list would not make sense and we could return back to array representation.